src/wallet/rpcwallet.cpp

```diff
@@ -4018,6 +4018,92 @@ UniValue walletcreatefundedpsbt(const JSONRPCRequest& request)
     return result;
 }
 
+UniValue upgradewallet(const JSONRPCRequest& request)
+{
+    std::shared_ptr<CWallet> const wallet = GetWalletForJSONRPCRequest(request);
+    CWallet* const pwallet = wallet.get();
+
+    if (!EnsureWalletIsAvailable(pwallet, request.fHelp)) {
+        return NullUniValue;
+    }
+
+    if (request.fHelp || request.params.size() > 1)
+        throw std::runtime_error(
+            RPCHelpMan{"upgradewallet",
+                "\nUpgrade the wallet. Upgrades to the latest version if no version number is specified\n"
+                "New keys may be generated and a new wallet backup will need to be made." +
+                    HelpRequiringPassphrase(pwallet) + "\n",
+                {
+                    {"version", RPCArg::Type::NUM, /* default */ strprintf("%d", FEATURE_LATEST), "The version number to upgrade to. Default is the latest wallet version"}
+                },
+                RPCResults{},
+                RPCExamples{
+                    HelpExampleCli("upgradewallet", "169900")
+                    + HelpExampleRpc("upgradewallet", "169900")
+                }
+            }.ToString());
+
+    RPCTypeCheck(request.params, {UniValue::VNUM}, true);
+
+    auto locked_chain = pwallet->chain().lock();
+    LOCK(pwallet->cs_wallet);
+
+    EnsureWalletIsUnlocked(pwallet);
+
+    int version = FEATURE_LATEST;
+    if (!request.params[0].isNull()) {
+        version = request.params[0].get_int();
+    }
+
+    int prev_version = pwallet->GetVersion();
+    if (version < prev_version) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Cannot downgrade wallet");
+    } else if (version == prev_version) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Wallet is already upgraded");
+    }
+
+    pwallet->SetMaxVersion(version);
+
+    pwallet->WalletLogPrintf("Performing wallet upgrade to %i\n", version);
+
+    // Do not upgrade versions to any version between HD_SPLIT and FEATURE_PRE_SPLIT_KEYPOOL unless already supporting HD_SPLIT
+    if (!pwallet->CanSupportFeature(FEATURE_HD_SPLIT) && version >= FEATURE_HD_SPLIT && version < FEATURE_PRE_SPLIT_KEYPOOL) {
+        throw JSONRPCError(RPC_WALLET_ERROR, "Cannot upgrade a non HD split wallet without upgrading to support pre split keypool. Please use the upgradewallet command.");
+    }
+
+    pwallet->SetMinVersion(version); // permanently upgrade the wallet after making sure we can upgrade
```

Comment:
This fails to compile.

Mind to make a move-only of the upgrade code into a standalone member function, that is merely called by the RPC instead of moving the whole wallet logic into the rpc file?

Reply:
Done